File: api/auth-signup.js
import { getUsersDb, hashPassword, createToken, sendJson, readBody } from './auth-shared.js';

export default async function handler(req, res) {
    if (req.method !== 'POST') return sendJson(res, 405, { error: 'Method not allowed' });

    const body     = await readBody(req);
    const username = body?.username?.trim();
    const email    = body?.email?.trim()?.toLowerCase() || null;
    const password = body?.password;

    if (!username)
        return sendJson(res, 400, { error: 'Username is required' });
    if (!password || String(password).length < 4)
        return sendJson(res, 400, { error: 'Password must be at least 4 characters' });

    try {
        const db    = await getUsersDb();
        const users = db.collection('users');

        // Ensure unique index exists
        await users.createIndex({ username: 1 }, { unique: true });

        const existing = await users.findOne({ username: username.toLowerCase() });
        if (existing) return sendJson(res, 409, { error: 'Username already exists. Try logging in!' });

        const doc = {
            username    : username.toLowerCase(),
            displayName : username,
            passwordHash: hashPassword(password),
            email       : email,
            createdAt   : new Date().toISOString(),
        };

        const result = await users.insertOne(doc);
        const token  = createToken({ userId: result.insertedId.toString(), username: doc.username });
        return sendJson(res, 201, { token, user: { id: result.insertedId.toString(), username: doc.displayName } });
    } catch (err) {
        if (err.code === 11000) return sendJson(res, 409, { error: 'Username already exists. Try logging in!' });
        console.error('[auth-signup]', err);
        return sendJson(res, 500, { error: err.message || 'Registration failed' });
    }
}

File: api/auth_signup.py
from datetime import datetime, timezone
from flask import Flask, request
from pymongo.errors import DuplicateKeyError
from api.auth_utils import get_users_db, hash_password, create_token, send_json

app = Flask(__name__)


@app.route('/api/auth/signup', methods=['POST'])
def signup():
    body = request.get_json(silent=True) or {}
    username = (body.get('username') or '').strip()
    email = (body.get('email') or '').strip().lower() or None
    password = body.get('password') or ''

    if not username:
        return send_json({'error': 'Username is required'}, 400)
    if not password or len(str(password)) < 4:
        return send_json({'error': 'Password must be at least 4 characters'}, 400)

    try:
        db = get_users_db()
        users = db.users
        users.create_index('username', unique=True)

        existing = users.find_one({'username': username.lower()})
        if existing:
            return send_json({'error': 'Username already exists. Try logging in!'}, 409)

        doc = {
            'username': username.lower(),
            'displayName': username,
            'passwordHash': hash_password(password),
            'email': email,
            'createdAt': datetime.now(timezone.utc).isoformat(),
        }

        result = users.insert_one(doc)
        token = create_token({
            'userId': str(result.inserted_id),
            'username': doc['username']
        })
        return send_json({
            'token': token,
            'user': {
                'id': str(result.inserted_id),
                'username': doc['displayName']
            }
        }, 201)
    except DuplicateKeyError:
        return send_json({'error': 'Username already exists. Try logging in!'}, 409)
    except Exception as e:
        print(f'[auth-signup] {e}')
        return send_json({'error': str(e) or 'Registration failed'}, 500)
